auditlog/proxy_decorator.go

```diff
@@ -0,0 +1,120 @@
+package auditlog
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ory/oathkeeper/driver/configuration"
+	"github.com/ory/oathkeeper/proxy"
+	"github.com/ory/x/logrusx"
+	"github.com/pkg/errors"
+)
+
+// RoundTripper interface is implemented by the Proxy structure and it's decorators.
+type RoundTripper interface {
+	RoundTrip(r *http.Request) (*http.Response, error)
+	Director(r *http.Request)
+}
+
+// ProxyAuditLogDecorator is a wrapper for Proxy struct with audit logging abilities.
+type ProxyAuditLogDecorator struct {
+	proxy    RoundTripper
+	builders []EventBuilder
+	senders  []Sender
+	logger   *logrusx.Logger
+}
+
+func NewProxyAuditLogDecoratorFromFile(proxy *proxy.Proxy, config configuration.Provider,
+	logger *logrusx.Logger) (*ProxyAuditLogDecorator, error) {
+	bs, err := DeserializeEventBuildersFromFiles(config.AuditLogConfigPath())
+	if err != nil {
+		return nil, err
+	}
+
+	return NewProxyAuditLogDecoratorFromEventBuilders(proxy, config, logger, bs)
+}
+
+func NewProxyAuditLogDecoratorFromEventBuilders(proxy *proxy.Proxy, config configuration.Provider,
+	logger *logrusx.Logger, bs []EventBuilder) (*ProxyAuditLogDecorator, error) {
+	d := &ProxyAuditLogDecorator{
+		proxy:    proxy,
+		builders: bs,
+		senders:  make([]Sender, 0),
+		logger:   logger,
+	}
+
+	d.senders = append(d.senders, &StdoutSender{l: logger})
+	if config.AuditLogKafkaEnabled() {
+		d.senders = append(d.senders, &KafkaSender{})
+	}
+
+	return d, nil
+}
+
+// RoundTrip performs wrapped structure's RoundTrip and logs request's event.
+func (d *ProxyAuditLogDecorator) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req == nil {
+		d.logger.Error("Request is nil")
+
+		return nil, errors.New("RoundTrip with nil request")
+	}
+
+	// Copy the request body before the request is sent further.
+	var reqBodyCopy io.ReadCloser
+	req.Body, reqBodyCopy = copyBody(req.Body, d.logger)
+
+	// Send request.
+	resp, err := d.proxy.RoundTrip(req)
+
+	// Deep copy request.
+	reqCopy := req.Clone(req.Context())
+	reqCopy.Body = reqBodyCopy
+
+	// Deep copy response.
+	respCopy := new(http.Response)
+	if resp != nil {
+		*respCopy = *resp
+		resp.Body, respCopy.Body = copyBody(resp.Body, d.logger)
```

Comment:
**[informational comment]**

Want to notify you that it can be tricky to manipulate the request/response body because every moment you should remember that in case of wrong actions you can face [goroutines leaks problem](https://husobee.github.io/golang/memory/leak/2016/02/11/go-mem-leak.html). So you should do it carefully, and it's also a good idea to have a Graphana chart with `goroutines number`, so you can check if you constantly create new goroutine and not finish them.

It's difficult for me to understand if you are facing this problem or not, so just be aware of this case.

Reply:
That's a great point!

What do you think about writing a stress test on that using something like [goleak](https://github.com/uber-go/goleak)?

Reply:
Hmm, according to the article, I should always close response bodies. I'll fix that.

Reply:
Well, adding `goleak` happened to be more complicated than I thought. 

It conflicts with some code in [ristretto](https://github.com/dgraph-io/ristretto) and finds additional goroutines sleeping after the test is ended. I'm not sure if it is an actual leak, most probably it is not.

Reply:
I think now I fixed it. Now all of the bodies are closed where it is necessary. But I can't verify it with a unit test.